### ISSUE TITLE
The future of modules

### ISSUE THREAD

**Issue by nilclass** [remotestorage/remotestorage.io#12]
TL;DR? This message is about changing the way modules work in remoteStorage.js. To skip the intro, search for "A) " in this message.

To recap, this is currently how modules work:
- Modules are defined through remoteStorage.defineModule. They consist of a name and a builder function.
- The name corresponds to a directory node directly below the storage root, as well as to an entry in the "scope" used to claim access.
- The builder function receives two BaseClient instances, corresponding to the private and public root of the module respectively.
- Modules return (amongst other things) an object called "exports", which is the public interface of the module, as seen by apps.
- Once a module is defined, it's "exports" object can be accessed via the remoteStorage object as remoteStorage[moduleName].

To illustrate this, here is what a module may look like:

``` javascript
remoteStorage.defineModule('bottles', function(privateClient, publicClient) {
  // declare which paths to use
  privateClient.use('');
  publicClient.use('');

  function generateId() {
    // (code ommitted)
  }

  return {
    exports: {
      add: function(content, size) {
        var id = generateId();
        return privateClient.storeObject('bottle', id, {
          content: content,
          size: size,
          id: id
        });
      },

      remove: privateClient.remove,

      getPublished: function() {
        return publicClient.getObject('publishedItems');
      },

      updatePublished: function(object) {
        publicClient.storeObject('published-items', 'publishedItems', object);
      },

      publish: function(id) {
        var published = this.getPublished();
        publicClient.storeObject('bottle', id, privateClient.getObject(id));
        published[id] = new Date().getTime();
        this.updatePublished(id);
      },

      unpublish: function(id) {
        var published = this.getPublished();
        publicClient.remove(id);
        delete published[id];
        this.updatePublished(id);
      }
    }
  }
});
```

This leads to a number of restrictions in writing apps:
1. In order to write an app, you need to write a module, unless you only depend on core-modules. This is even true for apps in which the module doesn't do anything that the BaseClient wouldn't be able to do on it's own.
2. The module determines the only public API that the app can use to access storage. This means that even very basic methods, such as getting / storing / deleting / listing objects or files, need to be implemented or explicitly delegated to the baseClient(s) by the module. Thus the structure and naming of the public API varies greatly between individual modules.
3. The separation of access to the "private" and "public" area through individual BaseClients makes it the module's responsibility to connect the two. Thus it becomes very hard to establish conventions for publishing and sharing, as those conventions cannot be implemented of the remoteStorage.js core code.
4. There is no defined way to extend a module. The public remoteStorage[moduleName] object can be extended like any other JavaScript object, but the syntax for this varies greatly from the regular module definition through defineModule().
5. Modules have no mechanism to access another person's storage.

In order to overcome these restrictions and reduce the amount of boilerplate code required to start with an app, I would like to make some changes.
Most of the following suggestions are the result of various discussions in the channel (#remotestorage on freenode) and at other places (such as issue #100). But before starting to actually implement those changes, I'd like to give everyone a chance to add to, criticize or approve with any of this. So go ahead :)
### A) Drop the public client, enhance the BaseClient

As can be seen in the example above, the code to publish objects is very generic. In fact, quite similar code is already part of the "root" module.
As publishing and sharing are the most common usecases for public data, it makes sense to have them as part of the BaseClient.

I propose to add the following methods to BaseClient:
- publishObject(path) - adds the given path with the current timestamp to the "publishedItems" object and copies the object found at "path" to public/{moduleName}/{path}
- shareObject(path, secret) - copies the object found at given path to public/{moduleName}/shared/{secret} and returns the absolute path to it as a String.
- getPublicClient(path) - retrieve a copy of the BaseClient that works on the "public" area. This would be for backwards compatibility with the current publicClient, as well as for custom operations done in the public area.
### B) Expose the BaseClient to the app

To make trivial access to remotestorage as simple as possible, it should be possible to gain access to a BaseClient instance without having to define a module.
This would work through a method called remoteStorage.getClient, which receives a moduleName:

``` javascript
var client = remoteStorage.getClient('bottles');
client.getObject('...');
client.publishObject('...');
```
### C) Make modules extensions of BaseClients instead of independent objects

Once (A) is implemented, our module boilerplate would have changed to:

``` javascript
defineModule('bottles', function(client) {
  return {
    exports: {}
  }
});
```

Then once (B) is implemented, all the generic methods from the module could disappear. All that remains is "add".
Now the situation is a bit weird, as some methods (such as remove, publishObject, ...) have to be called on a client, while "add" still has to be called on the module.

So instead of making assigning the "exports" object to remoteStorage[moduleName], the module could serve as an extension for the BaseClient.
The BaseClient in turn would be cached, so getClient('bottles') will return that modified BaseClient instead of a fresh one.

Thus once a module 'bottles' has been defined, the following lines become equivalent:

``` javascript
remoteStorage.bottles;
remoteStorage.getClient('bottles');
```

Note that in this case, the module would better name the "add" method "addBottle", to clearly distinguish it from adding other objects.

This would also enable modules to be extended, so a module that has been defined in the core, can be extended from an app by calling defineModule() again, without having to copy the entire module to the app.
### D) Apply modules to ForeignClient instances as well

Another new thing in the upcoming 0.7 release is the ForeignClient. It is basically a read-only BaseClient that is associated with another person's storage.
Once (C) has been implemented, the module's "exports" object can additionally extend any created "ForeignClient" for the same scope / moduleName.
To provide an example, consider a method "totalSize" for the example above. It would iterate over all "bottles" and add their "size" attributes. If the same method is applied to the ForeignClient, it can be used to perform the same operation on another persons published "bottles".

Ok, that's all for now. If I forgot anything, please add.

-- Niklas
